### ISSUE TITLE
Ignoring a component of loss while training

### ISSUE THREAD

**Issue by mjdietzx** [fchollet/keras#5045]
When training a GAN for example your discriminator may take a batch of images (either generated or real), and then multiple expected outputs for these images `[is_image_real, image_label]`.

In my case I have real images, some of which I have label's and others do not have labels. I want to use all of these images when training my GAN. This leads me to:

```python
disc.train_on_batch(image_batch_has_labels, [y_real, label_batch])
disc.train_on_batch(image_batch_no_labels, [y_real, no_labels_so_ignore_this_loss])
```
where `y_real = np.ones(shape=(batch_size, ))`.

The problem I am running into is how to specify the `no_labels_so_ignore_this_loss`. I don't want the discriminator to be penalized for predicting these labels since I don't have any ground truth label to evaluate against.

What is the most 'keras' way to accomplish this? The only thing I can really think of at this point is to somehow make the class weights 0 for the case of no labels.

**Comment by mjdietzx** [comment]
The way I am solving this right now is:

```python
disc.train_on_batch(image_batch_has_labels, [y_real, label_batch])

# since this batch of images has no labels, `label_batch` serves as a placeholder and it's values are garbage.
disc.train_on_batch(image_batch_no_labels, [y_real, label_batch], sample_weight=[None, np.zeros(batch_size)])
```

But I'm not sure this is a correct solution and haven't evaluated it if seems to get desired results yet.

**Comment by bstriner** [comment]
Are you trying to make a catGAN or something? Why does your discriminator predict labels as well? A standard discriminator input is an image and output is real/fake. A catGAN discriminator output is a set of categories. Are you trying to combine them?

If your problem decomposes, break the model up. Have a discriminator that predicts real/fake and a categorizer that predicts a category possibly based on some hidden layers of your discriminator. Then you just train the discriminator on all of the data and the categorizer on some of the data.

Another way of looking at it, just build a second model `disc2.train_on_batch(image_batch_no_labels, [y_real])` and it will share weights with the full model.

Cheers,
Ben

**Comment by mjdietzx** [comment]
What I'm implementing is similar to this (you can just look at the README to get an idea): https://github.com/buriburisuri/ac-gan. It looks like I'm having success with the sample_weights method I'm using although I want to verify that this makes sense.

Not only does my discriminator predict real/fake and the label of image, but it also predicts some of the noise that the generator used to generate the image. In the case of a real image there is no noise generating it so this loss should not be evaluated. I explained it with missing labels to simplify the problem since it is essentially the same thing.

I'd prefer not to have multiple models but what you suggested does sound like a valid solution. I just think a single model would perform better (takes a long time to fully train my GAN) and be cleaner.

**Comment by bstriner** [comment]
Let's say you create all your inputs and outputs using the functional API. You could create two models relatively easily, one that includes the extra target and one that doesn't. Both models share weights so you can train one a bit and train the other a bit. Just `model1=Model(x, [y,z])` and `model2=Model(x,[y,z,w])`.

The downside of using two models is that you're using two different optimizers so they don't share momentum. Shouldn't make a huge difference but you could hack around it if it bothers you.

The downside of setting weights to 0 is that during those iterations, the gradient due to the loss is 0, but you are still training, so the regularization loss will control everything. During the iterations when part of the network has 0 output weights, the parameters of that network will go to 0 due to regularization.

If you have two different models, the weights that aren't part of the model won't decay.

Haven't tried this model, so can't confirm yet, but worth testing. Please let me know how everything works out. I've been doing a ton of GANs recently:

https://github.com/bstriner/keras_adversarial

**Comment by bstriner** [comment]
By share weights, I mean they use the same layer instances, so everything is shared. 

You might also get better results if you mix batches instead of alternating.

You can build a model that does the supervised and the unsupervised training and run them simultaneously. For example, if you have some `model` with one input and two targets:
```python
x1 = Input(...)
x2 = Input(...)
m1 = model(x1)
m2 = model(x2)
mymodel = Model([x1, x2],[m1[0],m1[1],m2[0]])
```
So if your supervised data is x1 and the unsupervised data is x2, you could train batches including both.  Might be interesting to see if this makes training any better. You would want to run off equal batches of real and fake data, which is what you're probably doing anyways.

**Comment by mjdietzx** [comment]
OK, that makes sense. The solution you suggested will work for me I think, thanks! w.r.t. the regularization loss, are you referring to l1/l2 regularization? I see why that could cause trouble with my current solution of setting sample weights to 0. But if I'm not using l1/l2 reg is there anything I need to watch out for? I would expect things like dropout, batch normalization to be safe. I don't have great understanding of optimizers so not sure if this could mess up learning rates/momentum. Or where else this could have effects. If you know could you explain those cases?

But I do think this explains some weird behavior where losses looked strange in training and very different than when testing (where losses looked like I would expect them to).

As for the functionality of `sample_weights` I expected to specify sample(s) which have no effect on updating the weights but I guess this isn't the case? Is this the case by design or is it worth looking into implementing this?

**Comment by mjdietzx** [comment]
@bstriner your suggestion: 'Just model1=Model(x, [y,z]) and model2=Model(x,[y,z,w]).' is actually much cleaner to implement than what I had before. I'll let you know how the results look after some test runs.
